clients/tfchain-client-go/impl.go

```diff
@@ -37,145 +36,437 @@ type Versioned struct {
 type Conn = *gsrpc.SubstrateAPI
 type Meta = *types.Metadata
 
+// Pool connection
+type poolConn struct {
+	conn     Conn
+	meta     Meta
+	url      string
+	lastUsed atomic.Int64 // Unix timestamp
+	inUse    atomic.Bool
+}
+
+func (pc *poolConn) isHealthy() bool {
+	if pc == nil || pc.conn == nil || pc.meta == nil {
+		return false
+	}
+	_, err := getTime(pc.conn, pc.meta)
+	return err == nil
+}
+
+func (pc *poolConn) close() {
+	if pc != nil && pc.conn != nil {
+		pc.conn.Client.Close()
+		pc.conn = nil
+		pc.meta = nil
+		log.Debug().Str("url", pc.url).Msg("closed connection")
+	}
+}
+
 type Manager interface {
+	GetConnection(ctx context.Context) (*Substrate, error)
+	Close() error
+
+	// Deprecated methods
 	Raw() (Conn, Meta, error)
 	Substrate() (*Substrate, error)
 }
 
-type mgrImpl struct {
-	urls []string
+type ManagerConfig struct {
+	// Maximum number of connections in the pool
+	MaxPoolSize int
+	// Minimum number of connections to maintain
+	MinPoolSize int
+	// Maximum time a connection can be idle before being closed (if the pool has more than MinPoolSize)
+	MaxIdleTime time.Duration
+	// Interval between health checks
+	// After thinking about it, we don't need to periodically check the health of the connections
+	// because this creates a lot of overhead
+	// so instead we just check the health when we need to and do the maintanance in demand
+	// HealthCheckInterval time.Duration
+	// Timeout for creating new connections
+	ConnectionTimeout time.Duration
+}
+
+// Default configuration
+var DefaultConfig = ManagerConfig{
+	MaxPoolSize: 5,
+	MinPoolSize: 2,
+	MaxIdleTime: 30 * time.Minute,
+	// HealthCheckInterval: 120 * time.Second,
+	ConnectionTimeout: 10 * time.Second,
+}
+
+type manager struct {
+	urls      []string
+	pool      []*poolConn
+	mu        sync.RWMutex
+	ctx       context.Context
+	cancel    context.CancelFunc
+	wg        sync.WaitGroup
+	config    ManagerConfig
+	checkChan chan struct{}
+}
 
-	r int
-	m sync.Mutex
+func NewManager(urls ...string) Manager {
+	return NewManagerWithConfig(DefaultConfig, urls...)
 }
 
-func NewManager(url ...string) Manager {
-	if len(url) == 0 {
-		panic("at least one url is required")
+func NewManagerWithConfig(config ManagerConfig, urls ...string) Manager {
```

Comment:
This cannot be done in a backward-compatible way, can it?
To my understanding we will still have to maintain two functions `func NewManager(urls ...string) Manager` and `func NewManagerWithOptions(urls []string, opts ...Option) Manager`
I opted for having a second function to avoid breaking the old API